docs/transactions.md

```diff
@@ -0,0 +1,241 @@
+to a particular topic partition during the lifetime of a `single producer`. Transactional delivery allows producers to send data to multiple partitions such that either
+all messages are successfully delivered, or none of them are. Together, these capabilities enable `*exactly once semantics*` in Kafka.
+
+It is important to notice that:
+
+- `Transaction` always start from a `send` (producer)
+- Events sent to one or more topics will only be visible on consumers after the transaction is committed.
+- To use transactions, a `transaction id` (unique id per transaction) must be set prior an event is sent. If you do not provide one `kstreams` will auto generate one for you.
+- Transaction state is stored in a new internal topic `__transaction_state`. This topic is not created until the the first attempt to use a transactional request API. There are several settings to control the topic's configuration.
+- `Topics` which are included in transactions should be configured for durability. In particular, the `replication.factor` should be at least `3`, and the `min.insync.replicas` for these topics should be set to `2`
+- `Transactions` always add overhead, meaning that more effort is needed to produce events and the consumer have to apply filters
+For example, `transaction.state.log.min.isr` controls the minimum ISR for this topic.
+- `Streams` (consumers) will have to filter or not transactional events
+
+```plantuml
+@startuml
+!theme crt-amber
+
+!include https://raw.githubusercontent.com/plantuml-stdlib/C4-PlantUML/master/C4_Context.puml
+
+agent "Producer" as producer
+agent "Stream A" as stream_a
+agent "Stream B" as stream_b
+queue "Topic A" as topic_a
+
+producer -> topic_a: "produce with a transaction"
+topic_a --> stream_a: "consume only transactional events"
+topic_a --> stream_b: "consume all events"
+@enduml
+```
+
+## Transaction State
+
+A transaction have state, which is store in the special topic `__transaction_state`. This topic can be tuned in different ways, but it is outside of this guide scope.
+The only important detailt that you must be aware of is that a `transaction` can be `commited` or `aborted`. You will not manage the internal states but you are responsible
+of filtering or not transactions on the `stream` side.
+
+```plantuml
+@startuml
+!theme crt-amber
+
+!include https://raw.githubusercontent.com/plantuml-stdlib/C4-PlantUML/master/C4_Context.puml
+
+[*] --> Init
+Init : Initialized with a `transaction_id`
+
+Init -> Commited
+Commited --> [*]
+
+Init -> Aborted
+Aborted --> [*]
+
+@enduml
+```
+
+## Usage
+
+From the `kstreams` point of view, the `transaction pattern` is a `context manager` that will `start` a transaction and then `commit` or `aboort` it. Always a `transaction` starts when an event is send. Once that we have the `context` we can send events in two ways:
+
+Using the `StreamEngine` directly:
```

Comment:
This should be a side-note, the recommended way should be with the type annotation IMO.